### PR TITLE
ThirdPartyResource example: added watcher example, code cleanup

### PR DESCRIPTION
**Pull request by nilebox**
**What this PR does / why we need it**:
Added an example of using go-client for watching on ThirdPartyResource events (create/update/delete).
Watching TPRs is a very common case, but a rather complicated process (especially due to many non-trivial details like [client-go#8](https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502) and [kubernetes#16376](https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794))

Also, made minor code cleanup to avoid too many hardcoded strings copy-paste.

**Which issue this PR fixes**:
This PR gives an example of workarounds for [kubernetes#16376](https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794) and [kubernetes#26003](https://github.com/kubernetes/kubernetes/issues/26003#issuecomment-285674832)

### CODE REVIEW

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// newListWatchFromClient is a copy of cache.NewListWatchFromClient() method with custom codec
+// Cannot use cache.NewListWatchFromClient() because it uses global api.ParameterCodec which uses global
+// api.Scheme which does not know about custom types (Example in our case) group/version.
+// cache.NewListWatchFromClient(exampleClient, ExampleResourcePath, apiv1.NamespaceAll, fields.Everything())
+func newListWatchFromClient(c cache.Getter, resource string, namespace string, fieldSelector fields.Selector, paramCodec runtime.ParameterCodec) *cache.ListWatch {
```
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Would it make sense to raise another PR to include such method in client-go (as an alternative to cache.NewListWatchFromClient(), see the comment)?

**Comment by soltysh** [comment]
The code itself lgtm, but I'll let @sttts chime in.

**Comment by soltysh** [comment]
@k8s-bot ok to test

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		Group:   ExampleResourceGroup,
+		Version: ExampleResourceVersion,
+	}
+
+	schemeBuilder := runtime.NewSchemeBuilder(func(scheme *runtime.Scheme) error {
+		scheme.AddKnownTypes(
+			groupVersion,
+			&Example{},
+			&ExampleList{},
+			&metav1.ListOptions{},
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: Can we call `metav1.AddToGroupVersion(scheme, SchemeGroupVersion)` here instead of listing the `*Options` types one by one?

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+			&metav1.Status{},
+		)
+		return nil
+	})
+
+	scheme := runtime.NewScheme()
+	if err := schemeBuilder.AddToScheme(scheme); err != nil {
+		return nil, nil, err
+	}
+	if err := api.AddToScheme(scheme); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: Why do we add the core group?

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	config := *cfg
+	config.GroupVersion = &groupVersion
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(scheme)}
+
+	client, err := rest.RESTClientFor(&config)
+
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: nit: no empty line here

File: staging/src/k8s.io/client-go/examples/third-party-resources/types.go [hunk 5]
```diff
@@ -23,6 +23,18 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
+const (
+	ExampleDomain              = "k8s.io"
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/types.go]: it's not a good example to have resources under k8s.io. Can we move that to some sub-domain?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	store, controller := cache.NewInformer(
+		source,
+
+		// The object type.
+		&Example{},
+
+		// resyncPeriod
+		// Every resyncPeriod, all resources in the cache will retrigger events.
+		// Set to 0 to disable the resync.
+		0,
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Having some number `!=0` is good practice, isn't it?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Do we actually need our custom types in the `ParameterCodec`? It should only be used for `ListOptions` and friends.

**Comment by sttts** [comment]
Left a couple of comments.

File: staging/src/k8s.io/client-go/examples/third-party-resources/types.go (hunk 5 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/types.go]: This domain was there before, I just moved it to constant, but I agree.

> some sub-domain?

Do you mean something like "tpr.k8s.io" or completely different domain ("example.com" would clash with "example" resource there)? Any suggestions?

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go (hunk 3 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: This is a suggested workaround from there: https://github.com/kubernetes/kubernetes/issues/26003#issuecomment-285674832

Without this line, if you keep application with watcher running (in this example the application shuts down quickly, so you won't notice it, but if you just add a sleep for 10 minutes in the end, you'll see it), there is an error in the log occuring every 5 minutes:
```
2017/03/15 00:24:54 [WATCH] OnAdd /apis/tpr.k8s.io/v1/namespaces/default/examples/example1
E0315 00:29:54.550267   11041 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: no kind "Status" is registered for version "v1"
E0315 00:29:54.553813   11041 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: no kind "Status" is registered for version "v1"
2017/03/15 00:29:55 [WATCH] OnUpdate oldObj: /apis/tpr.k8s.io/v1/namespaces/default/examples/example1
2017/03/15 00:29:55 [WATCH] OnUpdate newObj: /apis/tpr.k8s.io/v1/namespaces/default/examples/example1
```

File: staging/src/k8s.io/client-go/examples/third-party-resources/types.go (hunk 5 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/types.go]: tpr.client-go.k8s.io would be nice.

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: So we need `Status` registered. Would prefer to only add that.

**Comment by nilebox** [comment]
@sttts fixed all the comments, please review. Once I get the **lgtm**, I will squash the commits.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 1 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: See https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// newListWatchFromClient is a copy of cache.NewListWatchFromClient() method with custom codec
+// Cannot use cache.NewListWatchFromClient() because it uses global api.ParameterCodec which uses global
+// api.Scheme which does not know about custom types (Example in our case) group/version.
+// cache.NewListWatchFromClient(exampleClient, ExampleResourcePath, apiv1.NamespaceAll, fields.Everything())
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: what's the purpose of this line?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 1 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Could you put the link into the comment?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 6 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: A resync event is going to be handled by OnAdd or OnUpdate, so it will be confusing. Plus in most cases a controller don't need to do resync. Hence, I prefer setting it to 0.

**Comment by caesarxuchao** [comment]
Thanks @nilebox! You'll also need to run hack/update-bazel.sh to get the Jenkin verification happy.

**Comment by nilebox** [comment]
@caesarxuchao thanks for the comments. I fixed the things you pointed at and squashed the commits.
e2e test is still failing, but looks like it is not related to my changes and fails in other branches as well.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go [hunk 8]
```diff
@@ -78,19 +77,29 @@ func main() {
...
+		clientset:     clientset,
+		exampleClient: exampleClient,
+		exampleScheme: exampleScheme,
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	go watcher.Run(ctx)
+
+	time.Sleep(5 * time.Second) // Wait until watcher starts
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: suspicious sleeps are suspicious. At least, add a comment saying that this sleep is not critical, only cosmetical (I hope it is!).

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Well, in the normal situation (not a hello world application) you wouldn't create an instance of TPR right after installing a watcher on it in the same app. So all this sleep is for is to wait until the watcher goroutine passes through the point of subscribing on updates. That's exactly what the comment is about.

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Which means that it is critical? I just want to avoid building cargo cults here because people copy this without understanding. So a proper 3 liner comment explaining this makes sense IMO.

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: And also, the normal app with watcher would keep running instead of exiting in 5-ish seconds

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Ok, I'll add a comment that in the normal app there is no need for sleep.

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: @sttts I've added the following comment:
```
	// The sleep below is just to make sure that the watcher.Run() goroutine has successfully executed
	// and the watcher is handling the events about Example TPR instances.
	// In the normal application there is no need for it, because:
	// 1. It's unlikely to create a watcher and a TPR instance at the same time in the same application.
	// 2. The application with watcher would most probably keep running instead of exiting right after the watcher startup.
```

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: 👍

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// See the issue comment: https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794
+// newListWatchFromClient is a copy of cache.NewListWatchFromClient() method with custom codec
+// Cannot use cache.NewListWatchFromClient() because it uses global api.ParameterCodec which uses global
+// api.Scheme which does not know about custom types (Example in our case) group/version.
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: As far as I see, the issue is not that the custom types `Example` etc. are not registered, but that `ListOptions` and friends must be registered under the `config.GroupVersion` of the client. Compare the registration in https://github.com/kubernetes/kubernetes/blob/db0b0bd0f5f934ef83cce0bd35212aba5bb31e18/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go#L46 and the conversion at https://github.com/kubernetes/kubernetes/blob/fb78390f839cd1241ef8cf8e4b257447001ba962/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L188.

In fact, I think we should do a proper comparison with all gvks at https://github.com/kubernetes/kubernetes/blob/fb78390f839cd1241ef8cf8e4b257447001ba962/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L187. After all the order of registration returned by `ObjectKinds` are arbitrary and uncontroller. @smarterclayton wdyt? Obviously, this is nothing for this PR.

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+		Group:   ExampleResourceGroup,
+		Version: ExampleResourceVersion,
+	}
+
+	schemeBuilder := runtime.NewSchemeBuilder(func(scheme *runtime.Scheme) error {
+		scheme.AddKnownTypes(
+			groupVersion,
+			&Example{},
+			&ExampleList{},
+		)
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: `&metav1.ListOptions{}` and `&metav1.DeleteOptions{}` should be added here otherwise get/list and delete will not work if this scheme object is used with client.

Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: Hm, interesting - do I understand it correctly that this is not necessary because of `metav1.AddToGroupVersion(scheme, groupVersion)`?

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @ash2k see 2 lines below:
```
metav1.AddToGroupVersion(scheme, groupVersion)
```
This invocation registers those (and other) types: https://sourcegraph.com/github.com/kubernetes/apimachinery@master/-/blob/pkg/apis/meta/v1/register.go#L39:6-39:23

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @ash2k see 2 lines below:
```
metav1.AddToGroupVersion(scheme, groupVersion)
```
This invocation registers those (and other) types: https://sourcegraph.com/github.com/kubernetes/apimachinery@master/-/blob/pkg/apis/meta/v1/register.go#L39:6-39:23

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		fmt.Printf("Failed to register watch for Example resource: %v\n", err)
+		return err
+	}
+
+	fmt.Printf("WaitForCacheSync\n")
+	// We must wait for exampleInformer to populate its cache to avoid reading from an empty cache
+	// in case of resource-generated evxents.
+	if !cache.WaitForCacheSync(ctx.Done(), exampleInformer.HasSynced) {
+		return errors.New("Wait for Example Informer was cancelled")
+	}
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Is this necessary here? Nothing works with cache/store in this goroutine so there is no point in blocking it before blocking on `<-ctx.Done()`.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+func (h *ExampleEventHandler) OnAdd(obj interface{}) {
+	example := obj.(*Example)
+	log.Printf("[WATCH] OnAdd %s\n", example.Metadata.SelfLink)
+}
+
+func (h *ExampleEventHandler) OnUpdate(oldObj, newObj interface{}) {
+	oldExample := oldObj.(*Example)
+	newExample := newObj.(*Example)
+	log.Printf("[WATCH] OnUpdate oldObj: %s\n", oldExample.Metadata.SelfLink)
+	log.Printf("[WATCH] OnUpdate newObj: %s\n", newExample.Metadata.SelfLink)
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Nit: `log.Print*()` does not need `\n` at the end, it will be added automatically. Also some places use `fmt.Print()` - this is inconsistent. Should probably do one or another.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 11 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Agreed, will remove this code.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 12 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: The initial example was using `fmt.Print()` so I wanted to use it. Missed `log.*` in some places, will fix.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 6 above)
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: From my limited experience I can say that in practice once every several minutes TCP connection times out if there were no events and is re-established. However things are happening in the cluster anyway and eventually history is lost and watch cannot be re-established from the last seen object (resourceVersion is too old and is no longer available). So an error is returned and informer will **have to** do the re-list. So having 0 here does not prevent resyncs, just makes them occur less often. Does this make sense?

Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: I mean, the consequence is that those callbacks have to handle "duplicate" update events anyway.

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: @ash2k not arguing with your experience, but I don't see how does it affect an example.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+		// resyncPeriod
+		// Every resyncPeriod, all resources in the cache will retrigger events.
+		// Set to 0 to disable the resync.
+		0,
+
+		// Your custom resource event handlers.
+		handler)
+
+	// store can be used to List and Get
+	// NEVER modify objects from the store. It's a read-only, local cache.
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Maybe mention that `exampleScheme.Copy()` can be used to make deep copies?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 6 above)
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: @nilebox it does not) Yeah, it is a bit off topic, sorry.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 13 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: Will do.

**Comment by nilebox** [comment]
@caesarxuchao @sttts anything else to be done to merge this pull request?

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+		scheme.AddUnversionedTypes(api.Unversioned, &metav1.Status{})
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: If this line comes from `api.addKnownTypes()`, shouldn't this line be
```
scheme.AddUnversionedTypes(api.Unversioned,
	&metav1.Status{},
	&metav1.APIVersions{},
	&metav1.APIGroupList{},
	&metav1.APIGroup{},
	&metav1.APIResourceList{},
)
```
Could someone who actually understands what is going on here tell us if this is the right thing to do?

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: The API* types are for discovery, not needed here.

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @ash2k adding `metav1.Status` type explicitly fixes the error:
```
E0310 11:28:15.006705   25312 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: no kind "Status" is registered for version "v1"
```
Using `scheme.AddUnversionedTypes` would help as well of course, but there is no need for extra types.

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: Adding `Status` here is definitely necessary. But we should move it to some central place like `metav1.AddToGroupVersion`. 

**But**, I start wondering whether we need this `Status` in all clients that do not  register the kube core group into their scheme. E.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1alpha1/register.go does not register `Status` as far as I see /cc @deads2k

Comment (deads2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: > But, I start wondering whether we need this Status in all clients that do not register the kube core group into their scheme. E.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1alpha1/register.go does not register Status as far as I see /cc @deads2k

It's needed for proper serving and I supposed for proper client receipt, but the types themselves didn't require it so I add to the scheme later.  I have no objection to moving where the registration happens though.

Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: Do we still need the concept of "Unversioned"? I think they are replaced by "meta/v1" now. I think add "Status" to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go#L46 is enough.

**Comment by caesarxuchao** [comment]
@nilebox i left a comment about moving registering Status to `metav1.AddToGroupVersion`, otherwise lgtm.

File: staging/src/k8s.io/client-go/examples/third-party-resources/client.go (hunk 14 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @caesarxuchao moving this line to `metav1.AddToGroupVersion()` leads to circular dependency between `metav1` and `api` packages 😞 

Why don't just invoke `api.addKnownTypes()` which already registers `Status` then?

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @nilebox the `Status` type is in meta already. Only the Group-Version constant is not. The circular dependency should be easy so remove.

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client.go]: @nilebox we shouldn't add a dependency from every client to `pkg/api`. Third-party clients might be independent from the core API. We don't want to have all core types in every client's scheme.

**Comment by nilebox** [comment]
@caesarxuchao Done, `Status` registration is moved to `metav1.AddToGroupVersion()`.

Is it safe to replace `Unversioned` with `metav1.SchemeGroupVersion` and get rid of `Unversioned` completely? (see [kubernetes](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/register.go#L127) and [client-go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/pkg/api/register.go#L127))
If so, I can raise a separate PR for this.

**Comment by nilebox** [comment]
The build test fails due to overriden `apiVersion` by registering Status in `metav1.AddToGroupVersion()` (with custom apiVersion different from `Unversioned`), if I understood correctly.
@caesarxuchao Should I update tests, or fix the code somehow?

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go (hunk 13 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: As far as I see, we do not have deepcopy functions generated for the example types. This means `scheme.Copy` will use reflection. This is slow, not well tested and **will go away soon**.

File: staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+	// You can use exampleScheme.Copy() to make a deep copy of original object and modify this copy
+	for _, obj := range store.List() {
+		example := obj.(*Example)
+		exampleScheme.Copy(example)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go]: even though I discourage the use of the reflection based deepcopy code path, this line is a noop anyway. `example, err = scheme.Copy(example)` would make sense.

**Comment by nilebox** [comment]
I will raise a new PR after resolving conflicts and cleanup.
